### ISSUE TITLE
Step marked as "Phase: Succeeded" has failed exit code and parent marked as failed

### ISSUE THREAD

**Issue by ajcann** [argoproj/argo#3374]
**What happened**:

When a pod is evicted due to spot revocation, the pod message is "failed with exit code 137" or "pod termination" but in either case the node phase is marked as "Succeeded". The parent node's phase then becomes "Failed" with "child '[name of child]' failed".

**What you expected to happen**:

The child node phase should be 'Error' and the retryStrategy which is 'OnError' should cause the child to retry.

**How to reproduce it (as minimally and precisely as possible)**:

Cordon and drain a node with a running child node pod to mimic a spot eviction.

- Argo version:
```
v2.8.2
```
- Kubernetes version :
```
1.15
```

```
DEBU[0000] CLI version                                   version="{2.8.2+8a151ae.dirty 2020-06-18T23:50:58Z 8a151aec6538c9442cf2380c2544ba3efb60ff60 2.8.2 dirty go1.13 gc linux/amd64}"
DEBU[0000] Client options                                opts="{{ false false} 0x1574670 0xc0000ff9f0}"
Name:                work-gluc26zpc
Namespace:           prod
ServiceAccount:      default
Status:              Failed
Conditions:
 Completed           True
Created:             Wed Jul 01 12:01:34 -0700 (1 hour ago)
Started:             Wed Jul 01 12:01:34 -0700 (1 hour ago)
Finished:            Wed Jul 01 13:35:17 -0700 (19 minutes ago)
Duration:            1 hour 33 minutes

STEP                                                                                                                TEMPLATE       PODNAME                     DURATION  MESSAGE
 ✖ work-gluc26zpc                                                                                                  batchsets-run
 ├-✔ batchsets(0:batchset-uris:s3://redact-project-bucket/work-glucl-hm-bilayer1_mul/lbe_r4/run0-wf0-job0.json)  batchset
 | ├---✔ sample(0)                                                                                                  sample         work-gluc26zpc-1687359342  59m
 | └---✔ score(0)                                                                                                   score          work-gluc26zpc-2158749979  29m
 ├-✖ batchsets(1:batchset-uris:s3://redact-project-bucket/work-glucl-hm-bilayer1_mul/lbe_r4/run0-wf0-job1.json)  batchset                                             child 'work-gluc26zpc-2578364283' failed
 | └---✔ sample(0)                                                                                                  sample         work-gluc26zpc-3049864182  19m       pod termination
 ├-✔ batchsets(2:batchset-uris:s3://redact-project-bucket/work-glucl-hm-bilayer1_mul/lbe_r4/run0-wf0-job2.json)  batchset
 | ├---✔ sample(0)                                                                                                  sample         work-gluc26zpc-4184041754  1h
 | └---✔ score(0)                                                                                                   score          work-gluc26zpc-742636919   28m
 ├-✖ batchsets(3:batchset-uris:s3://redact-project-bucket/work-glucl-hm-bilayer1_mul/lbe_r4/run0-wf0-job3.json)  batchset                                             child 'work-gluc26zpc-556078815' failed
 | └---✔ sample(0)                                                                                                  sample         work-gluc26zpc-1342035298  19m       failed with exit code 137
 ├-✔ batchsets(4:batchset-uris:s3://redact-project-bucket/work-glucl-hm-bilayer1_mul/lbe_r4/run0-wf0-job4.json)  batchset
 | ├---✔ sample(0)                                                                                                  sample         work-gluc26zpc-2985116534  1h
 | └---✔ score(0)                                                                                                   score          work-gluc26zpc-2484502147  28m
 └-✖ batchsets(5:batchset-uris:s3://redact-project-bucket/work-glucl-hm-bilayer1_mul/lbe_r4/run0-wf0-job5.json)  batchset                                             child 'work-gluc26zpc-1083581027' failed
   └---✔ sample(0)                                                                                                  sample         work-gluc26zpc-645921838   19m       pod termination
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by sarabala1979** [comment]
@ajcann I am not able to reproduce the issue with the below examples.  Let me know  if I am missing anything

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: loops-param-result-
spec:
  entrypoint: loop-param-result-example
  templates:
    - name: loop-param-result-example
      steps:
        - - name: generate
            template: gen-number-list
        - - name: failedWith137
            template: retry-script
            arguments:
              parameters:
                - name: seconds
                  value: "{{item}}"
            withParam: "{{steps.generate.outputs.result}}"

    - name: gen-number-list
      script:
        image: python:alpine3.6
        command: [python]
        source: |
          import json
          import sys
          json.dump([i for i in range(0, 5)], sys.stdout)

    - name: retry-script
      retryStrategy:
        limit: 1
      inputs:
        parameters:
          - name: seconds
      script:
        image: python:alpine3.6
        command: ["python"]
        # fail with a 66% probability
        source: |
          import random;
          import sys;
          sys.exit(137)
```


```
Name:                loops-param-result-7qmjc
Namespace:           default
ServiceAccount:      default
Status:              Failed
Message:             child 'loops-param-result-7qmjc-1943968283' failed
Conditions:          
 Completed           True
Created:             Thu Jul 02 08:24:33 -0700 (27 seconds ago)
Started:             Thu Jul 02 08:24:33 -0700 (27 seconds ago)
Finished:            Thu Jul 02 08:25:00 -0700 (now)
Duration:            27 seconds
ResourcesDuration:   0s*(100Mi memory),35s*(1 cpu)

STEP                                TEMPLATE                   PODNAME                              DURATION  MESSAGE
 ✖ loops-param-result-7qmjc         loop-param-result-example                                                 child 'loops-param-result-7qmjc-1943968283' failed  
 ├---✔ generate                     gen-number-list            loops-param-result-7qmjc-3722579240  3s                                                            
 └-·-✖ failedWith137(0:0)           retry-script                                                              No more retries left                                
   | ├-✖ failedWith137(0:0)(0)      retry-script               loops-param-result-7qmjc-1934723350  5s        failed with exit code 137                           
   | └-✖ failedWith137(0:0)(1)      retry-script               loops-param-result-7qmjc-1330581971  8s        failed with exit code 137                           
   ├-✖ failedWith137(1:1)           retry-script                                                              No more retries left                                
   | ├-✖ failedWith137(1:1)(0)      retry-script               loops-param-result-7qmjc-1080099970  3s        failed with exit code 137                           
   | └-✖ failedWith137(1:1)(1)      retry-script               loops-param-result-7qmjc-475958591   5s        failed with exit code 137                           
   ├-✖ failedWith137(2:2)           retry-script                                                              No more retries left                                
   | ├-✖ failedWith137(2:2)(0)      retry-script               loops-param-result-7qmjc-2204940510  7s        failed with exit code 137                           
   | └-✖ failedWith137(2:2)(1)      retry-script               loops-param-result-7qmjc-527031515   6s        failed with exit code 137                           
   ├-✖ failedWith137(3:3)           retry-script                                                              No more retries left                                
   | ├-✖ failedWith137(3:3)(0)      retry-script               loops-param-result-7qmjc-3012936882  6s        failed with exit code 137                           
   | └-✖ failedWith137(3:3)(1)      retry-script               loops-param-result-7qmjc-3482563119  6s        failed with exit code 137                           
   └-✖ failedWith137(4:4)           retry-script                                                              No more retries left                                
     ├-✖ failedWith137(4:4)(0)      retry-script               loops-param-result-7qmjc-3787507662  5s        failed with exit code 137                           
     └-✖ failedWith137(4:4)(1)      retry-script               loops-param-result-7qmjc-3183366283  7s        failed with exit code 137        
```

**Comment by ajcann** [comment]
It's turns out this issue is a side effect of an underlying Kubernetes issue https://github.com/kubernetes/kubernetes/issues/58711 This has been patched in Kubernetes as of 1.17.11, 1.18.8,  1.19.0. I do not believe it has been ported back to 1.16 or earlier.
